### PR TITLE
[k8s] Share SSH jump pod

### LINKED ISSUE #2598: [k8s] Share SSH jump pod across users in the same namespace
We currently create a SSH jump pod per user, even if all of them use the same namespace:

https://github.com/skypilot-org/skypilot/blob/b1d8c9d8bfca9d9d696ba3ef07a2f2cdf91b061a/sky/clouds/kubernetes.py#L31

This should instead by based on the kubecontext hash (like we have in `get_current_user_identity`) so that different users sharing the same namespace should use the same SSH jump pod. 

We _may_ also need additionally handle keys and auth from different users on the same ssh jump pod.

### PR DESCRIPTION
**Pull request by kbrgl**
This PR closes #2598 and adds plumbing to share the SSH jump pod across a Kubernetes namespace.

1. Per-user secrets for the user's public SSH key are replaced with a single secret object, `sky-ssh-key`, that is created once and patched when future attempts to setup Kubernetes authentication are made.
2. For initial jump pod creation, the manifest runs `cat /etc/secret-volume/ssh-key-* > ~/.ssh/authorized_keys` to set up the initial user's SSH key.
3. When a new user tries to authenticate, the patched secret is re-mounted at /etc/secret-volume and this change is picked up by the LCM script which reruns `cat /etc/secret-volume/ssh-key-* > ~/.ssh/authorized_keys`.

The LCM script now uses 2 threads to poll: one to manage the lifecycle and terminate the jump pod, and one to reload keys on an interval.

## Testing

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_compatibility_tests.sh`

Tested locally.

cc @romilbhardwaj

### CODE REVIEW

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py [hunk 1]
```diff
@@ -3,15 +3,20 @@
...
-kubeapi), if it does not see ray pods in the duration of 10 minutes. If the
+kubeapi) if it does not see ray pods in the duration of 10 minutes. If the
 user re-launches a task before the duration is over, then ssh jump pod is being
-reused and will terminate itself when it sees that no ray cluster exist in that
-duration.
+reused and will terminate itself when it sees that no ray clusters exist in
+that duration.
+
+This script also reloads SSH keys from the mounted secret volume on an
+interval.
```
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:12]:
```suggestion
To allow multiple users to the share the same SSH jump pod, 
this script also reloads SSH keys from the mounted secret volume on an
interval and updates `~/.ssh/authorized_keys`.
```

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py [hunk 2]
```diff
@@ -29,82 +34,135 @@
...
-        time.sleep(retry_interval)
-
-        # List the pods in the current namespace
+def poll(interval, leading=True):
+    """Decorator factory for polling function. To stop polling, return True.
+    
+    Args:
+        interval (int): The amount of time to wait between function calls.
+        leading (bool): Whether to wait before (rather than after) calls.
+    """
```
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:51]: Nice!

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py [hunk 3]
```diff
@@ -29,82 +34,135 @@
...
+        sys.stdout.write('[SSH Key Reloader] Terminated.\n')
+        return True
+
+    # Reload SSH keys from mounted secret volume
+    sys.stdout.write('[SSH Key Reloader] Reloading SSH keys.\n')
+    cmd = 'cat /etc/secret-volume/ssh-key-* > ~/.ssh/authorized_keys'
+    try:
+        subprocess.check_output(cmd, shell=True)
+    except Exception as e:
+        sys.stdout.write(f'[SSH Key Reloader] [ERROR] failed to reload SSH keys: {e}\n')
```
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:85]: Nit
```suggestion
        sys.stdout.write(f'[SSH Key Reloader][ERROR] Failed to reload SSH keys: {e}\n')
```

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py [hunk 4]
```diff
@@ -29,82 +34,135 @@
...
+            'service.\n')
         try:
-            ret = v1.list_namespaced_pod(current_namespace,
-                                         label_selector=label_selector)
+            # ssh jump resources created under same name
+            v1.delete_namespaced_service(current_name, current_namespace)
+            v1.delete_namespaced_pod(current_name, current_namespace)
         except Exception as e:
-            sys.stdout.write(f'Error: listing pods failed with error: {e}\n')
+            sys.stdout.write('[Lifecycle] [ERROR] Deletion failed. Exiting '
```
Comment (romilbhardwaj) [inline sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:134]: Nit
```suggestion
            sys.stdout.write('[Lifecycle][ERROR] Deletion failed. Exiting '
```

**Review by romilbhardwaj — COMMENTED** [review]
Awesome, thanks @kbrgl! Took a skim through the code and it looks good. Left some nits. Will test it and approve if all is good.

File: sky/clouds/kubernetes.py [hunk 5]
```diff
@@ -27,8 +27,9 @@
 class Kubernetes(clouds.Cloud):
     """Kubernetes."""
 
-    SKY_SSH_KEY_SECRET_NAME = f'sky-ssh-{common_utils.get_user_hash()}'
-    SKY_SSH_JUMP_NAME = f'sky-ssh-jump-{common_utils.get_user_hash()}'
+    SKY_SSH_KEY_SECRET_NAME = 'sky-ssh-key'
```
[the remark this answers is not in the document]
Comment (kbrgl) [reply sky/clouds/kubernetes.py:30]: Yeah I think that's a good idea!

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py (hunk 2 above)
Comment (kbrgl) [reply sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:52]: 🥳

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @kbrgl! Did some manual testing - seems to be working nicely. Backward compatibility has some issues, see comments.

File: sky/utils/kubernetes/ssh_jump_lifecycle_manager.py [hunk 6]
```diff
@@ -29,82 +35,135 @@
...
+def reload_keys():
+    """Reloads SSH keys from mounted secret volume."""
+
+    if terminated:
+        sys.stdout.write('[SSH Key Reloader] Terminated.\n')
+        return True
+
+    # Reload SSH keys from mounted secret volume
+    sys.stdout.write('[SSH Key Reloader] Reloading SSH keys.\n')
+    cmd = 'cat /etc/secret-volume/ssh-key-* > ~/.ssh/authorized_keys'
```
[the remark this answers is not in the document]
Comment (romilbhardwaj) [reply sky/utils/kubernetes/ssh_jump_lifecycle_manager.py:82]: On a related note, I've opened #2901 to find a better solution to this backward compatibility problem.

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @kbrgl! Pushed some nits and merge conflict fixes. Manually tested backward compat, now running smoke tests.

Once everything is ready, will merge this PR and push new container images to our image registry.

**Review by romilbhardwaj — APPROVED** [review]
Thanks @kbrgl! Smoke tests pass. Manually tested backward compatibility. Building and updating our docker image, will merge after that's done.

**Comment by romilbhardwaj** [comment]
* Updated images and build script to use date versioning - https://github.com/skypilot-org/skypilot-catalog/pull/60
* Fixed #3158 
* Added socat and netcat to base image on user request for faster provisioning

- [ ] Running k8s smoke tests

**Pull request merged**